tools/autograd/gen_variable_type.py

```
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+
+        input_num = 0
+        bwd_name = name + '_backward'
+        assert len(declarations_by_name[bwd_name]) == 1
+        bwd = declarations_by_name[bwd_name][0]
+
+        def actual(arg):
             name = arg['name']
-            sizes_name = name + '_sizes'
-            if any(re.search(name_regex.format(name), f) for f in derivatives):
-                saved.append(arg)
-            if any(sizes_name in f for f in derivatives):
-                saved.append({
-                    'name': sizes_name,
-                    'type': 'IntList',
-                })
-            for f in derivatives:
-                for match_name in re.findall(r"{}_argsize_\w+".format(name), f):
-                    saved.append({
-                        'name': match_name,
-                        'type': 'int64_t',
-                    })
-                for match_name in re.findall(r"{}_argsizes_\w+".format(name), f):
-                    saved.append({
-                        'name': match_name,
-                        'type': 'IntList',
-                    })
-        option['saved'] = saved
-
-        options.append(option)
-
-    options = sorted(options, key=lambda o: o['name'])
-    for name, overloads in groupby(options, lambda o: o['name']):
-        overloads = list(overloads)
-        for i, option in enumerate(overloads):
-            name = option['name']
-            option['op'] = name[0].upper() + name[1:] + 'Backward'
-            if len(overloads) > 1:
-                option['op'] += str(i)
-
-    return options
-
-
-def create_autograd_functions(top_env, declarations):
+            return name if name != 'inplace' else 'false'
+
+        actuals = [actual(arg) for arg in bwd['arguments']]
+        formula = '{}({})'.format(bwd_name, ', '.join(actuals))
+        formula = formula.replace('grad_output', 'grad')
+        if not re.search(IDENT_REGEX.format('grad'), formula):
+            formula = '({}).mul_(grad)'.format(formula)
```

Uhhh this makes me nervous. Maybe let's just raise an error if `grad` doesn't appear anywhere

---

Yes, but why can't we include that in the formula itself? Plus `mul_` with 2 tensors is going to raise an error in backward

---

I'm not sure what you mean. These four lines are constructing the formula (from the signature of the backwards function) for the first derivative of NN functions.  The formulas aren't specified anywhere else: it's implicit from the naming of xxx_forwards and xxx_backwards 

(The second derivatives are handled in derivatives.yaml).

---

Ohhhhh I see. Makes sense now. Thank you